File: interview-questions-kotlin/src/test/kotlin/dlind1974/algo/top/FizzBuzzTest.kt
package dlind1974.algo.top

import kotlin.test.Test
import kotlin.test.assertEquals

private fun fizzBuzz(n: Int): List<String> {
    var result = mutableListOf<String>()

    for (i in 1..n) {
        if (i % 15 == 0) result.add("FizzBuzz")
        else if (i % 3 == 0) result.add("Fizz")
        else if (i % 5 == 0) result.add("Buzz")
        else result.add(i.toString())
    }

    return result
}

private fun fizzBuzzIdiomatic(n: Int): List<String> = List<String>(n) { i ->
    val index = i + 1
    when {
        index % 15 == 0 -> "FizzBuzz"
        index % 3 == 0 -> "Fizz"
        index % 5 == 0 -> "Buzz"
        else -> index.toString()
    }
}

class FizzBuzzTest {

    @Test
    fun testN3() {
        val actual = fizzBuzzIdiomatic(3)
        val expected = listOf("1","2","Fizz")
        assertEquals(expected, actual)
    }

    @Test
    fun testN5() {
        val actual = fizzBuzzIdiomatic(5)
        val expected = listOf("1","2","Fizz","4","Buzz")
        assertEquals(expected, actual)
    }

    @Test
    fun testN15() {
        val actual = fizzBuzzIdiomatic(15)
        val expected = listOf("1","2","Fizz","4","Buzz","Fizz","7","8","Fizz","Buzz","11","Fizz","13","14","FizzBuzz")
        assertEquals(expected, actual)
    }
}

File: interview-questions-cpp/top-interview-questions/FizzBuzzTest.cpp
#include <gtest/gtest.h>
#include <string>
#include <vector>

using namespace std;

namespace {
vector<string> fizzBuzz(int n) {
  vector<string> result;
  for (auto i = 1; i <= n; ++i) {
    if (i % 15 == 0) {
      result.emplace_back("FizzBuzz");
    } else if (i % 3 == 0) {
      result.emplace_back("Fizz");
    } else if (i % 5 == 0) {
      result.emplace_back("Buzz");
    } else {
      result.emplace_back(to_string(i));
    }
  }
  return result;
}
} // namespace

TEST(FizzBuzz, Test3) {
  auto actual = fizzBuzz(3);
  vector<string> expected = {"1", "2", "Fizz"};
  ASSERT_EQ(expected, actual);
}

TEST(FizzBuzz, Test5) {
  auto actual = fizzBuzz(5);
  vector<string> expected = {"1", "2", "Fizz", "4", "Buzz"};
  ASSERT_EQ(expected, actual);
}

TEST(FizzBuzz, Test15) {
  auto actual = fizzBuzz(15);
  vector<string> expected = {"1",    "2",    "Fizz", "4",    "Buzz",
                             "Fizz", "7",    "8",    "Fizz", "Buzz",
                             "11",   "Fizz", "13",   "14",   "FizzBuzz"};
  ASSERT_EQ(expected, actual);
}